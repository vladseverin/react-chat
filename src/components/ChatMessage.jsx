import React from 'react';
import { withStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import moment from 'moment';
import classNames from 'classnames';
import Avatar from './Avatar';

import getColor from '../utils/color-form';
import senderName from '../utils/sender-name';

const styles = theme => ({
  messageWrapper: {
    display: 'flex',
    justifyContent: 'flex-start',
    alignItems: 'center',
    padding: `${theme.spacing.unit}px ${theme.spacing.unit * 3}px`,
  },
  messageWrapperFromMe: {
    justifyContent: 'flex-end',
  },
  message: {
    maxWidth: '70%',
    minWidth: '10%',
    padding: theme.spacing.unit,
    marginLeft: theme.spacing.unit * 2,
  },
  messageFromMe: {
    marginLeft: 0,
    marginRight: theme.spacing.unit * 2,
    backgroundColor: '#e6dcff',
    order: -1,
  },
  statusMessage: {
    textAlign: 'center',
    marginLeft: 'auto',
    marginRight: 'auto',
  },
});

const ChatMessage = ({
  classes, content, sender, activeUser, createdAt, statusMessage,
}) => {
  const displayedName = senderName(sender);

  const isMessageFromMe = sender._id === activeUser._id;

  const userAvatar = (
    <Avatar colorFrom={sender._id}>
      {displayedName}
    </Avatar>
  );

  if (statusMessage) {
    return (
      <div className={classes.messageWrapper}>
        <Typography className={classes.statusMessage}>
          <Typography variant="caption" style={{ color: getColor(sender._id) }} className={classes.statusMessageUser}>
            {displayedName}
          </Typography>
          {content}
          <Typography variant="caption" component="span">
            {moment(createdAt).fromNow()}
          </Typography>
        </Typography>
      </div>
    );
  }

  return (
    <div className={isMessageFromMe
      ? classNames(classes.messageWrapperFromMe, classes.messageWrapper)
      : classes.messageWrapper}
    >
      {userAvatar}
      <Paper className={isMessageFromMe
        ? classNames(classes.messageFromMe, classes.message)
        : classes.message}
      >
        <Typography
          variant="caption"
          style={{ color: getColor(sender._id) }}
        >
          {displayedName}
        </Typography>
        <Typography variant="body1" >
          {content}
        </Typography>
        <Typography variant="caption">
          {moment(createdAt).fromNow()}
        </Typography>
      </Paper>
    </div>
  );
};

export default withStyles(styles)(ChatMessage);
