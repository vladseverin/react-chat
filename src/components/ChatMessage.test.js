/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import ChatMessage from './ChatMessage';

jest.mock('moment', () => () => ({ fromNow: () => 'a few seconds ago' }));

const mockProps = {
  content: 'Hey, Wassup',
  sender: {
    _id: '12345',
    firstName: 'Vladislav',
    lastName: 'Severin',
    username: 'vlad',
  },
  activeUser: {
    _id: '12345',
  },
  createdAt: '2018-06-13T20:48:03.200Z',
  statusMessage: false,
};

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<ChatMessage {...mockProps} />, div);
  ReactDOM.unmountComponentAtNode(div);
});
