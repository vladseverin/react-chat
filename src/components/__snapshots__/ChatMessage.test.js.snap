// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatMessage /> renders correctly 1`] = `
<div
  className="ChatMessage-messageWrapperFromMe-61 ChatMessage-messageWrapper-60"
>
  <div
    className="MuiAvatar-root-65 MuiAvatar-colorDefault-66"
    style={
      Object {
        "backgroundColor": "#d32f2f",
      }
    }
  >
    VS
  </div>
  <div
    className="MuiPaper-root-68 MuiPaper-elevation2-72 MuiPaper-rounded-69 ChatMessage-messageFromMe-63 ChatMessage-message-62"
  >
    <span
      className="MuiTypography-root-95 MuiTypography-caption-105"
      style={
        Object {
          "color": "#d32f2f",
        }
      }
    >
      Vladislav Severin
    </span>
    <p
      className="MuiTypography-root-95 MuiTypography-body1-104"
    >
      Hey, Wassup
    </p>
    <span
      className="MuiTypography-root-95 MuiTypography-caption-105"
    >
      a few seconds ago
    </span>
  </div>
</div>
`;
